docs/orleans/implementation/cluster-management.md

```diff
@@ -36,148 +38,143 @@ In addition to the <xref:Orleans.IMembershipTable> each silo participates in a f
 
-1. For the implementation of this feature we utilize the support for transactions over multiple rows that are provided by the `MembershipTable`.
-1. We add a membership-version row to the table that tracks table changes.
-1. When silo S wants to write suspicion or death declaration for silo P:
+1. [Apache ZooKeeper](https://zookeeper.apache.org/) - in this implementation we use the configured deployment ID as a root node and the silo identity (`ip:port@epoch`) as its child node. Together they guarantee a unique path per silo. For concurrency control, we use optimistic concurrency control based on the [node version](https://zookeeper.apache.org/doc/r3.4.6/zookeeperOver.html#Nodes+and+ephemeral+nodes). Every time we read from the deployment root node we store the version for every read child silo node and use that version when we try to write back. Each time a node's data changes, the version number increases atomically by the ZooKeeper service. For multi-row transactions, we utilize the [multi method](https://zookeeper.apache.org/doc/r3.4.6/api/org/apache/zookeeper/ZooKeeper.html#multi(java.lang.Iterable)), which guarantees serializable transactions over silo nodes with the same parent deployment ID node.
 
-    1. S reads the latest table content. If P is already dead, do nothing. Otherwise,
-    1. In the same transaction, write the changes to P's row as well as increment the version number and write it back to the table.
-    1. Both writes are conditioned with ETags.
-    1. If the transaction aborts due to ETag mismatch on either P's row or the version row, attempt again.
+1. [Consul IO](https://www.consul.io) - we used [Consul's Key/Value store](https://www.consul.io/intro/getting-started/kv.html) to implement the membership table. Refer to [Consul-Deployment](../deployment/consul-deployment.md) for more details.
 
-1. All writes to the table modify and increment the version row. That way all writes to the table is serialized (via serializing the updates to the version row) and since silos only increment the version number, the writes are also totally ordered in increasing order.
+1. [AWS DynamoDB](https://aws.amazon.com/dynamodb/) - In this implementation, we use the cluster Deployment ID as the Partition Key and  Silo Identity (`ip-port-generation`) as the RangeKey making the record unity. The optimistic concurrency is made by the `ETag` attribute by making conditional writes on DynamoDB. The implementation logic is quite similar to Azure Table Storage.
 
-**Scalability of the extended membership protocol:**
+1. In-memory emulation for development setup. We use a special system grain for that implementation. This grain lives on a designated primary silo, which is only used for a **development setup**. In any real production usage primary silo **is not required**.
```

Comment:
Entry for Cassandra provider:

1. [Apache Cassandra](https://cassandra.apache.org/) - in this implementation we use the composite of Service Id and Cluster Id as partition key and the silo identity (`ip:port:epoch`) as row key. Together they guarantee a unique row per silo. For concurrency control, we use optimistic concurrency control based on a [static column](https://cassandra.apache.org/doc/latest/cassandra/reference/static.html) `version` using a [Lightweight Transaction](https://cassandra.apache.org/doc/4.1/cassandra/architecture/guarantees.html#lightweight-transactions-with-linearizable-consistency). This `version` column is shared for all rows in the partition/cluster so provides the consistent incrementing version number for each cluster's membership table. There are no multi-row transactions in this implementation.